### PR TITLE
hmem/cuda: avoid stub loading at runtime

### PR DESCRIPTION
**Pull request by aws-nslick**
When the CUDA toolkit is installed, a set of "stub" libraries are installed under /usr/local/cuda*/lib64/stubs/. These libraries include a SONAME field with a `.1' suffix, but the filenames of these stubs are bare. eg:

 > $ readelf -d /usr/local/cuda-12.5/lib64/stubs/libnvidia-ml.so | grep soname
 > 0x000000000000000e (SONAME)  Library soname: [libnvidia-ml.so.1]

The CUDA toolkit does not include any library file with the name `libnvidia-ml.so.1` (or `libcuda.so.1`, etc.), as these are provided by the driver package. This disconnect between the stub filename in the toolkit and the SONAME within it is done intentionally to allow linking with the stub at build time, while ensuring it's never loaded at runtime.

In normal dynamic linking cases (ie: without dlopen), the SONAME field of `libnvidia-ml.so.1` is used in the DT_NEEDED tag, where that filename can only come from a driver package and this ensures that the stub library will never match.

Match the same behavior and provide `.1` suffixes to dlopen where appropriate for NVIDIA libraries.

### CODE REVIEW

**Comment by aws-nslick** [comment]
bot:aws:retest

**Review by shijin-aws — APPROVED** [review]
Can you rename the PR/commit title to be `hmem/cuda: avoid stub loading at runtime`

**Comment by aws-nslick** [comment]
> Can you rename the PR/commit title to be hmem/cuda: avoid stub loading at runtime

Done, thanks. 

One other thing I'd ask reviewers to think about is to consider searching inside `CUDA_HOME/lib64/stubs` in autoconf, instead of today where it searches only in `CUDA_HOME/lib64` and `CUDA_HOME/lib`, which would make it possible to build libfabric without any build-time dependency on the driver; the cuda toolkit alone would be sufficient for all cases.

**Review by j-xiong — APPROVED** [review]

**Comment by aws-nslick** [comment]
Please wait on an ack from @bwbarrett before merging as there was some disagreement on a similar change [here](https://github.com/aws/aws-ofi-nccl/pull/581) and I want to make sure we're aligned.

**Comment by bwbarrett** [comment]
Note that I objected to this commit in the OFI plugin because it doesn't match what Nvidia has done in the past with NCCL.  If we want to make a change, it should be to adopt `cudaGetDriverEntryPoint()`.

**Comment by aws-nslick** [comment]
@bwbarrett such changes were made on the ofi plugin side; do you have a problem with this approach for libfabric?

File: fabtests/common/hmem_cuda.c
Comment (aws-nslick) [reply fabtests/common/hmem_cuda.c]: I removed anything related to libcudart -- this should be restricted to just libcuda.so.1 and libnvidia-ml.so.1

**Review by shijin-aws — APPROVED** [review]

**Review by j-xiong — APPROVED** [review]

**Pull request merged**
